Create chapter 2 content (#36)
joelostblom/viz-oer: textbook/2_grouping.qmd

---
title: "Grouping"
include-in-header:
  - text: |
      <style>
        .table.dataframe {
          width: auto;
          text-align: right;
        }
      </style>
fig-width: 4
fig-height: 4
resources:
  - data
  - utils.py
---

::: {.callout-note title="Learning outcomes"}
- **Explain** how grouping allows you to directly compare categories and helps you to avoid making incorrectly generalized conclusions (e.g., Simpson’s Paradox).

> - **Apply** grouping strategies using color, size, shape and position to visually distinguish between categories via `.encode(color='category')` and `aes(color=category)`.

Reviewer: One thought I had for LOs in general, is that how do we include code if more than two packages are used in the book... although that can be a future problem as I'm not sure how likely it is to happen and I don't think it is worth removing  the code just for that reason

- **Apply** the `facet` method to split data into separate subplots, revealing patterns in each group.

- **Select** the grouping strategy that best highlights your key comparison. Use color to focus on specific values, and apply faceting to reveal broader trends across categories.

:::

## The case for grouping

 So far you have created declarative visualizations by representing them as 'marks' on a canvas with encoded x- and y-axis positions. But this approach doesn't always tell the whole story. Real-world data is inherently messy, and trends that emerge in one context may evaporate in another. 

Consider the following example:

::::: {.callout-default .ex-prompt}

In Viridia, historians have been studying datasets from Earth to identify what went wrong and avoid repeating the mistakes of their ancestors.

> During their research, they uncovered records of a toxic beverage known as “alcohol.” 

Author: Great idea. I have tried to implement this idea in the latest commit by hiding the chart in a hint and asking:

"Try to imagine what this chart could look like. What trend would you expect to see in this data?"

They discovered a glorious correlation: countries with higher life expectancies often have a greater percentage of heavy alcohol drinkers. They now plan to introduce alcohol into Viridia.

Below is a chart illustrating their findings:

```{python}
#| label: fig-polar

> #| fig-cap: "Source: Global Burden of Disease Study 2021 (GBD 2021) Results; Human Mortality Database (2023), United Nations - World Population Prospects (2022) – processed by Our World in Data"

Author: I tried using footnotes for now as you can hover over them and see the source in place. I should probably add the url too so they can directly download the csv.

import altair as alt
import pandas as pd

life_expectancy_df = pd.read_csv("data/life_expectancy_alcohol.csv")

alt.Chart(life_expectancy_df).mark_circle().encode(

>     x=alt.X('average_life_expectancy', scale=alt.Scale(zero=False)),

Author: Yes I agree, it looks quite weird without fitting to data range. I've switched it out for a dataset showing life expectancy from age 50 which doesn't require using Alt and has the same effect. I tried a bunch of other datasets for this example (deaths, cancer, heart disease, car crashes etc) and it's the best I could find.

As we have seen with alcohol consumption and life expectancy, grouping has major benefits. It allows us to:

1. **Directly compare** trends and values for distinct groups

2. **Discover trends** in the data that were previously obscured

3. **Avoid making inaccurate conclusions** about the relationships in our data

To understand these benefits, let's consider Simpson's Paradox. The paradox was not invented by Homer Simpson et al., but rather Edward H. Simpson, a British codebreaker. 

Simpson reported a curious phenomenon where a trend present in several different groups of data disappears when they are combined. 

Below is an illustration of the Simpson's Paradox at play. Notice how the overall dataset does not have a clear direction but when you color by group, both groups exhibit the same negative trend. 

::: {.panel-tabset}

### Overall

```{python}
simpsons_df = pd.read_csv("data/simpsons_paradox_data_l.csv")
alt.Chart(simpsons_df).mark_circle().encode(
  x=alt.X('X'),
  y=alt.Y('Y'),
)

> ```

Reviewer: Do we need to switch to a toy dataset here? Could we continue with the same data set, a smaller version of that data, or a related data set?

### By group

```{python}
alt.Chart(simpsons_df).mark_circle().encode(
  color='Group',
  x=alt.X('X'),
  y=alt.Y('Y'),
)
```

:::

In this case, grouping makes the insidious Simpson's Paradox visually unmistakable. This is a prime example of the dual purpose of data visualization. It doesn't just present our data to others; it deepens our understanding of it. 

## Strategies for grouping

> The grammar of graphics gives us a convenient syntax for grouping. We simply define the encoding channel that should represent a categorical variable, using `aes()` in ggplot and `.encode()` in Altair. Color, shape, size and position are the most effective channels for grouping, and they can be easily interchanged:

Reviewer: Let's try to keep all mentions of library specific (such as `aes` and `encode`) inside the panels of the respective package

::: {.panel-tabset}

# Altair
::: {.panel-tabset}

# Color

```python
.encode(color = 'category_column')
```

# Shape

```python
.encode(shape = 'category_column')
```

# Size

```python
.encode(size = 'category_column')
```

# Position

```python
.encode(x = 'category_column')
```

```python
.encode(y = 'category_column')
```

> :::

Reviewer: No spaces around parameter assignment in python

# ggplot
::: {.panel-tabset}
# Color

```python
aes(color = category_column)
```

# Shape

```python
aes(shape = category_column)
```

# Size

```python
aes(size = category_column)
```

# Position

```python
aes(x = category_column)
```

```python
aes(y = category_column)
```
:::
:::

The main challenge in applying grouping is understanding which encoding channel to use to facilitate the comparisons we want to make. Each grouping strategy has unique strengths and pitfalls, which we shall discuss in detail.

> ### Color

Reviewer: For some reason, this doesn't quite  match what I see in the rendered version. Maybe the render didn't update after you made another commit? 

![image](https://github.com/user-attachments/assets/9f90eee0-e9ad-485b-9cbd-a405dfa7ee18)

::: {.panel-tabset}

# Altair
```{python}
alt.Chart(simpsons_df).mark_circle().encode(
  color='Group',
  x=alt.X('X'),
  y=alt.Y('Y'),
)
```

# ggplot

```{r}
# Load necessary libraries
library(ggplot2)
simpsons_df <- read.csv("data/simpsons_paradox_data_l.csv")

ggplot(simpsons_df, aes(x = X, y = Y, color = Group)) +
  geom_point()
```

:::

It is critical that each color be clearly distinct from the others, as too-similar colors can confuse the viewer and make differentiation difficult. Similarly, it is wise to limit the number of distinct categories represented by color: adding too many different colors overwhelms the visualization and makes them harder to distinguish.

Accessibility is another key concern when using color. Colorblindness affects approximately 8% of men and 0.5% of women globally, with the most common variant being deuteranopia, which impairs the ability to distinguish between red and green hues. While Altair’s default color palette is designed to be color-blind friendly, ggplot’s is not.

> 

Reviewer: ```suggestion

We will learn much more about colors later in chapter ...including specifics around which color are safe to use together and which are suitable to represent different data types.
```

Author: have added this

### Shape

Shape can be used strategically in combination with color to provide an additional visual cue. This is particularly effective in contexts where colors may be difficult to distinguish. 

A classic example of the power of this combination is the humble traffic light. At first glance, traffic lights seem poorly designed: they use the worst possible colors for people the most common form of colorblindess: red and green. However, in many cities, traffic lights are encoded with using both color and shape. The combination of these encoding channels, alongside the use of predictable y-axis positions for both 'Stop' and 'Go' signals creates a powerful effect that has undoubtedly prevented many accidents.

![](/img/traffic.png)

We can encode shapes aesthetics the same way as color, this can make it easier to distinguish the groups. 

::: {.panel-tabset}

# Altair
```{python}
alt.Chart(simpsons_df).mark_circle().encode(
  shape='Group',
  color='Group',
  x=alt.X('X'),

>   y=alt.Y('Y'),

Reviewer: No need to use alt.X/Y when not passing options

)
```

# ggplot

```{r}
# Load necessary libraries
library(ggplot2)

> simpsons_df <- read.csv("data/simpsons_paradox_data_l.csv")

Reviewer: let's use `readr::read_csv` over `read.csv`, there are some cases where the latter causes issues that are hard to detect. In general, I wonder if we should read from the public URL here instead of the relative path, so that student can easily copy the code and run on their computers?

Author: have changed!

ggplot(simpsons_df, aes(x = X, y = Y, shape = Group, color=Group)) +
  geom_point()
```

:::

> The combination of shape and color can make grouping clearer than either of these categories alone. However, there are some important caveats. At line-drawings or outlines can make color difficult to distinguish as it reduces the area that is colored. It is preferable to use filled shapes when combining these encoding channels. 

Reviewer: What does "At line-drawings" mean?

Author: have corrected

> Another important consideration is the size of the shapes. It is much easier for the viewer to distinguish between larger shapes than smaller ones. If you have a scatterplot with tiny marks, you may want to restrict the number of unique shapes you show, or consider avoiding shape altogether.

Reviewer: It would be good if we can give heuristics for the number of shapes. I'm not sure if the rules are the same as color here, my sense is that it might be even lower

### Size

> Size is more useful for ordinal categories. Hello.

Reviewer: hi

:::: {.callout-default .ex-hint}

Alcohol consumption is influenced by income. How could we slightly alter the chart to show this?

::: {.callout-default .ex-solution}

The simplest solution is to color each country by its World Bank income group.

```{python}

alt.Chart(income_df).mark_point().encode(
    color='income_group',
    x='percentage_heavy_drinkers',
    y='life_expectancy_at_50',
)
```

Notice how the data points form clusters by income group. Coloring reveals that both national alcoholism levels and life expectancy are correlated with income. <span style="color: #9CCCC8; font-weight: bold">High income</span> countries tend to have longer lifespans and a greater proportion of heavy alcohol drinkers. 

We see the opposite picture for most <span style="color: #5D84A8; font-weight: bold">low income</span>, where both indicators are lower. 

That could explain our mysterious correlation!

By grouping the data, we have revealed that a **confounding variable**: income influences both life expectancy and alcohol consumption. This is sobering news for Viridia...

:::

::::

::::

> :::::

Author: @joelostblom I like this it feels like a proper story. Also makes sense why you are in such high demand in all of these Viridian departments!

2. **Discover patterns** in the data that were previously obscured

3. **Avoid making inaccurate conclusions** about the relationships in our data

To understand these benefits, let's consider Simpson's Paradox. The paradox was not invented by Homer Simpson et al., but rather Edward H. Simpson, a British codebreaker. 

Simpson reported a curious phenomenon where a trend present in several different groups of data disappears when they are combined. 

Below is an illustration of the Simpson's Paradox at play. Notice how the overall dataset does not have a clear direction but when you color by group, both groups exhibit the same negative trend. 

::: {.panel-tabset}

### Overall

```{python}
simpsons_df = pd.read_csv("data/simpsons_paradox_data_l.csv")
alt.Chart(simpsons_df).mark_point().encode(
  x='X',
  y='Y',
)
```

### By group

```{python}
alt.Chart(simpsons_df).mark_point().encode(
  color='Group',
  x='X',
  y='Y',
)
```

:::

In this case, grouping makes the insidious Simpson's Paradox visually unmistakable. This is a prime example of the dual purpose of data visualization. It doesn't just present our data to others; it deepens our understanding of it. 

> 

Author: That makes sense! I did this and moved the text "While position and color are generally the most effective grouping strategies, it is important to understand the unique strengths and pitfalls of each strategy to chose the one that best supports the comparisons we want to make. This is what we will learn about next" to before the grouping strategies section.

```
:::

# ggplot
::: {.panel-tabset}
# Color

```python
aes(color=category_column)
```

# Shape

```python
aes(shape=category_column)
```

# Size

```python
aes(size=category_column)
```

# Position

```python
aes(x=category_column)
```

```python
aes(y=category_column)
```
:::
:::

The main challenge in applying grouping is understanding which encoding channel best supports the comparisons we want to make. Each grouping strategy has unique strengths and pitfalls, which we shall discuss in detail.

### Color

The human brain can detect color differences with little cognitive effort. Color is a **pre-attentive feature** which means that the brain can tell apart colors even without conscious attention. This is heightened when there is a high contrast, such as between the complimentary colors blue and orange.

This is great news for grouping! By encoding our categorical variable as color, we can leverage the brain's ability to infer groups. For scatterplots, this works best with fewer than eight categories. 

Conveniently, color doesn't alter the layout or structure of the data visualization, so it's an efficient way to compare the exact values of data points between groups.

> Let's use color to group our life expectancy data by world region:

Author: Have added the ref!

> Let's use color to group our life expectancy data by world region:

Reviewer: Intriguing question setup in the paper!

> In some cases, including that search for circles among squares, the reaction time (RT) to declare that the target was present or to declare that all the items were the same did not vary as a function of how many items were presented. The slope of the RT x set size function was near zero. This strongly suggested that the all the items were being processed “in parallel” [[2](https://pmc.ncbi.nlm.nih.gov/articles/PMC6513732/#R2)]. In other cases, the RT increased as a roughly linear function of set size, suggesting a serial process [though that need not be the case [3](https://pmc.ncbi.nlm.nih.gov/articles/PMC6513732/#R3),[4](https://pmc.ncbi.nlm.nih.gov/articles/PMC6513732/#R4)]. What made some targets “pop-out”?

# Altair

Here we use `color='region'`:

```{python}
alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='life_expectancy_at_50',
    color='region',
)
```

# ggplot

Here we use `color=region`:
```{r fig.width=8, fig.height=6}
library(readr)
library(ggplot2)
income_df <- read_csv('data/income_lifeexp_alcohol_1.csv', show_col_types = FALSE)

ggplot(income_df, aes(
    x=percentage_heavy_drinkers, 
    y=life_expectancy_at_50, 
    color=region
)) + 
geom_point()
```

:::

Notice how you naturally interpret points of same color as clusters?

But beware... The power of color fades when used irresponsibly. Using   colors that are too similar can blur the difference between groups. The more colors you include, the harder it is to stick to distinct hues.

There is also the challenge of accessibility. Colorblindness affects approximately 8% of men and 0.5% of women globally. The most common variant is deuteranopia. This impairs the ability to distinguish between red and green hues, so designing your data visualization with this in mind is essential. While Altair’s default color palette is designed to be color-blind friendly, ggplot’s is not.

> We will learn much more about accessibility in chapter ...including specifics around which colors are safe to use together and which are suitable to represent different data types.

Author: @joelostblom Which comment do you mean?

### Shape

Now our visualizations are really taking shape! Shape can be combined with color to provide an additional visual cue. Let's take a look at a classic example: the infamous traffic light. 

> Do you think that traffic lights are well designed? At first glance,  they break a major rules of accessibility, using the worst possible colors for people with deuteranopia (red and green). Thankfully, in many countries, traffic lights are encoded using both color and shape. A third encoding channel - position - is then used to establish predictable y-axis positions for both 'Stop' and 'Go' signals. 

Author: @joelostblom  page 101 of Visualization Analysis & Design by Tamara Munzner could be a good shout. It has the hierarchy of categorical encoding channels, where shape is actually bottom.

![](/img/traffic.png)

Grouping by three encoding channels reinforces the visual contrast and has probably prevented many accidents.

We can use the same syntax to encode both color and shape aesthetics: 

::: {.panel-tabset}

# Altair

Here we use `shape='region'`:

```{python}
alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='life_expectancy_at_50',
    shape='region',
    color='region',
)
```

# ggplot

Here we use `shape=region`:

```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x=percentage_heavy_drinkers, 
    y=life_expectancy_at_50, 
    shape=region,
    color=region
)) + 
geom_point()
```

:::

While this combined approach makes grouping clearer, there are some pitfalls. For instance, line icons limit the total area that can be colored, reducing the impact of color. It is better to use solid shapes when combining shape and color.

> Additionally, size matters more than you'd think. Smaller shapes are difficult to distinguish. When plotting a dense scatterplot of small marks, try restricting the number of unique shapes to three or fewer, or consider avoiding using shape altogether.

Author: @joelostblom  I think I made this rule up... Visualization Analysis & Design by Tamara Munzner says that with 10x10 pixels you can make around 12 distinguishable shapes. Grouping by 12 shapes seems it could be a spaghetti nightmare

# Create the scatter plot
ggplot(income_df_subset, aes(
    x=percentage_heavy_drinkers, 
    y=life_expectancy_at_50, 
    size=income_group,
)) + 
geom_point()

```

:::

::: {.callout-caution}
In general, size-based groupings are not very intuitive. Restricting size to four or fewer levels is a good rule of thumb, and for most visualizations grouping by size should be avoided.
:::

### Position

Position is a powerful grouping strategy. It works by encoding our categorical variable as the x or y axis. Each group has its own visual lane, making it easier to compare overall distributions across categories. It  also emphasizes the outliers within each category.

Below, we group on the y-axis to compare heavy alcohol consumption across world regions:

::: {.panel-tabset}

# Altair

Here we use `y='region'`:

```{python}
alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='region',
)
```

# ggplot

Here we use `y=region`:

```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x=percentage_heavy_drinkers, 
    y=region, 
)) + 
geom_point()
```

:::

Notice how easy it is to compare the distributions between regions?

> We can even take it one step further. Position can be combined with color to reinforce grouping on one variable or to introduce grouping on a second variable. 

Author: Yep I like this idea. I've moved the section below jittering. The chart is a lot more interpretable with jittering, otherwise it seems kinda pointless to group by color.

When grouping by two categorical variables, which should you assign to position and which to color? Position is more visually obvious than color. This means that we assign our primary comparison to position and our secondary comparison to color. 

Let's apply this to a real world example.

::::: {.callout-default .ex-prompt}

> You are trying to understand how life expectancy varies across income groups globally, and the role world region plays in these differences.

Author: I tried to do this subtly without distracting from the main question

#|   - facet_alt
import altair as alt
import pandas as pd
from utils import assert_chart_equal

url = 'data/income_lifeexp_alcohol_1.csv'

```

```{pyodide}
#| exercise: position_alt
income_df = pd.read_csv(url)

# Write code above to create the chart 
```

::: {.callout-default .ex-solution}

```{python}
alt.Chart(income_df).mark_point().encode(
    x='life_expectancy_at_50',
    y='income_group',
    color='region',
)
```

:::

# ggplot

```{webr fig.width=8, fig.height=6}
#| setup: true
#| exercise:
#|   - position_ggplot
#|   - facet_ggplot
library(ggplot2)

url <- 'data/income_lifeexp_alcohol_1.csv'
```

```{webr fig.width=8, fig.height=6}
#| exercise: position_ggplot
income_df <- readr::read_csv(url, show_col_types=FALSE)

# Write code above to create the chart 
```

::: {.callout-default .ex-solution}

You are primarily comparing the `income_group` variable, so it should be grouped by position on either the `x` or `y` axis. The secondary comparison, `region`, should be encoded as color. 
```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x=life_expectancy_at_50, 
    y=income_group, 
    color=region
)) + 
geom_point()
```

:::

:::

:::::

> For all of these benefits, there are some drawbacks of grouping by position. For one, it occupies an axis that could have represented a continuous variable. This reduces the information density of your visualization. More worryingly, visually separating each group by position can hide overall trends in the dataset. It also requires significant space.

Author: Yes I meant the latter.

Have changed to:
More worryingly, while other grouping methods, such as color, allow you visualize both overall and group-level trends,  overall trends are difficult to see when grouping by position

Lastly, plotting data points along a single axis can lead to  overlap. That's where **jittering** comes to the rescue. Jittering introduces slight variation, or 'jitter', in the position of data points in each lane, making them easier to see. 

::: {.column-margin}

**Jitterbug**

Back in the 1930s, the jitterbug was a lively, often wild, dance style that matched the fast-paced, unpredictable energy of swing music. In data visualization, jittering serves a similar purpose: to add a touch of randomness to points in a plot.

![](/img/640px-Jitterbug_Wolcott_FSA.jpg)

:::

Let's examine how jittering works.

::: {.panel-tabset}

# Altair

In Altair, we create a random calculated variable called `jitter` by appending the `.transform_calculate()` method to our chart. We use our newly created `jitter` variable, encoding it as the `yOffset` of each data point using `yOffset='jitter:Q'`. This shifts the `yOffset` of each point by a random value. 

> The `:Q` simply tells Altair that `jitter` is a continuous quantitative variable.

Author: Agreed! it seems like a basic functionality that is more difficult than it needs to be

```{python}
alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='region',
    color='region',
    yOffset="jitter:Q"
).transform_calculate(
    jitter="random()"
)
```

# ggplot

> In ggplot2, you can achieve the same effect with jittering by using the `position_jitter()` function. Jitter is directly applied to the marks through the position argument within `geom_point()`.
> 
> ```{r fig.width=8, fig.height=6}
> ggplot(income_df, aes(
>     x=percentage_heavy_drinkers, 
>     y=region, 
>     color=region
> )) + 
> geom_point(position = position_jitter(width = 0, height = 0.2))

Reviewer: ```suggestion
In ggplot2, you can jitter points by using the `geom_jitter`, which is a combination of `geom_point` together with a function called `position_jitter`.

```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x=percentage_heavy_drinkers, 
    y=region, 
    color=region
)) + 
geom_jitter()
```

Was there a specific reason you opted for position_jitter here? Maybe because it is more explicit? I see geom_jitter more commonly, but happy to switch if you think the other approach is more instructive.

```

:::

::: {.callout-caution}
Jittering can be confusing! It can easily be mistaken for a scatterplot with true variation in the underlying data.

> :::

Reviewer: Hmm is this true? Maybe something to discuss.

Author: Maybe this is just my problem. Take that chart I shared before. For me, on first glance it looks like the y-axis position of the points has some sort of significance or meaning. 

![image18-1-1024x645](https://github.com/user-attachments/assets/aa3ad4bd-f3c9-4caf-a25c-9cd5f5c692a5)

> ## Faceting

Reviewer: ```suggestion
### Faceting
```

I think this should still be part of the "grouping strategies" section

> Faceting takes the concept of grouping to the next level by generating multiple small charts, each dedicated to a subset of the data. It is particularly effective for highlighting distinct trends and patterns in each subset, as well as identifying extreme values or outliers within groups. And unlike grouping by position, faceting keeps both the x and y axis free for other variables.
> 
> Let's check out the syntax:

Reviewer: ```suggestion
Faceting takes the concept of grouping to the next level. Instead of using different visual elements within a single chart, it generates multiple small charts, each dedicated to a grouped subset of the data. Faceting is particularly effective for highlighting distinct trends and patterns in each subset, as well as identifying extreme values or outliers within groups. We could think of faceting as a type of position grouping, but its advantage is that it keeps both the x and y axis free for other variables.

Let's check out the syntax:
```

::: {.panel-tabset}

# Altair

> In Altair, faceting by World Bank income group is straightforward. We simply append `.facet('income_group')` to our chart. Additionally, we can control the layout by setting the number of columns using the `columns` argument:

Reviewer: ```suggestion
To facet by income group, we can append `.facet('income_group')` to our chart. When we are faceting the xy-axes are the same for all the grouped facets/subcharts. The default layout in Altair is to have single row with all the charts spread out horizontally. This would make it easy to compare the charts' y-axis, since they would all be aligned. If we instead are the most interested in comparing the values on the x-axis, we could spread the charts out vertically in a single column. Here, we don't have an explicit primary comparison, but we show how to change from the default horizontal layout by using the `columns` parameter.
```

```{python}
alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='life_expectancy_at_50',
).facet(
    'income_group',

>     columns=2

Reviewer: ```suggestion
    columns=1
```

)

> ```

Reviewer: ````suggestion
```

In addition to faceting on a single variable in the data, faceting can also be done over two variables: one spread horizontally and one spread out vertically in the faceted charts. In Altair, we would use the `column` and `row` parameter in order to achieve this. For example, to have facets that are grouped by both income group and region, we could use `facet(column='income_group', row='region')`. This can be effective when we need to group our data further, but it does risk making the chart hard to overview, so it is best used when there are fewer categories in the faceting variables.
````

# ggplot

> Facetting works via `facet_wrap`. By default it tries to create an even num of cols and rows, rather than putting all plot in one row like Altair. However, you can specify the desired number of columns with the `ncol` argument.

Reviewer: ```suggestion
To facet by income group, we can append `.facet_wrap(vars(income_group))` to our chart. The `vars` function is required for `facet_wrap` to understand that we are referring to column names just like `aes` is required when typing out column names for the different visual channels.

When we are faceting the xy-axes are the same for all the grouped facets/subcharts. The default layout in ggplot is to achieve and even number of columns and rows. This is good when we are low on space, but otherwise it is preferred to spread the charts out horizontally in a single row if the primary purpose was to compare the charts' y-axis, since they would all be aligned, or vertically in a single column if we are the most interested in comparing the groups in terms of the values on the x-axis. Here, we don't have an explicit primary comparison, but we show how to change from the default horizontal layout by using the `ncol` parameter.
```

`vars` is recommend over ~ in general nowadays, neither is very intuitive in my opionion.

```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x = percentage_heavy_drinkers, 
    y = life_expectancy_at_50
)) +
geom_point() +

> facet_wrap(~ income_group, ncol=2)

Reviewer: ````suggestion
facet_wrap(~ income_group, ncol=1)
```

In addition to faceting on a single variable in the data, faceting can also be done over two variables: one spread horizontally and one spread out vertically in the faceted charts. In ggplot, we would swap out `facet_wrap` to `facet_grid` in order to do this. For example, to have facets that are grouped by both income group and region, we could use `facet_grid(vars(income_group), vars(region))`. This can be effective when we need to group our data further, but it does risk making the chart hard to overview, so it is best used when there are fewer categories in the faceting variables.
````

```

:::

> Faceting makes it easy to see the overall shape of the distribution for each income group and how tightly clustered the values are. It also reveals outliers in the Upper Middle Income facet – we can easily spot two countries with a higher percentage of heavy drinkers.

Reviewer: ```suggestion
As you can see above, faceting makes it easy to observse the overall shape of the distribution for each income group and how tightly clustered the values are. It also reveals potential outliers in the Upper Middle Income facet – we can easily spot two countries with a higher percentage of heavy drinkers. On the other hand, it is now harder to compare the exact values between two groups because the points are spread out in different charts, instead of right next to each other in a single chart.
```

> Another advantage is that we can apply additional layers of grouping within each facet. Now its your turn to try this!

Reviewer: ```suggestion
All the previous grouping strategies that we have discussed can be utilized within each facet as additional layers of grouping. This nested grouping can be powerful to stratify our data according to multiple categorical variables, but keep in mind that it can become confusing if we try to fit too much into a single visualization. Exactly how many levels of grouping you can use depends on how  well separated your data is and how many observations you have; more observations and less separation between groups means that you will have to be more careful and use less levels of grouping to keep the visualization easy to interpret.

Let's try out an example of how we can use faceting together with other grouping strategies.
```

::::: {.callout-default .ex-prompt}

Edit this chart to use both faceting and grouping.
Facet by world region (`region`) and group by World Bank income group (`income_group`). 
Ensure that only two facets are shown per row.

::: {.panel-tabset}

# Altair

```{pyodide}
#| exercise: facet_alt
income_df = pd.read_csv(url)

alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='life_expectancy_at_50',
)

```

# ggplot

```{webr fig.width=8, fig.height=6}
#| exercise: facet_ggplot
income_df <- readr::read_csv(url, show_col_types=FALSE)

ggplot(income_df, aes(
    x=percentage_heavy_drinkers, 
    y=life_expectancy_at_50, 
)) + 
geom_point()
```

:::

::: {.callout-default .ex-solution}

::: {.panel-tabset}

# Altair

```{python}
alt.Chart(income_df).mark_point().encode(
    x='percentage_heavy_drinkers',
    y='life_expectancy_at_50',
    color='income_group'
).facet(
    'region',
    columns=2
)
```

# ggplot

```{r fig.width=8, fig.height=6}
ggplot(income_df, aes(
    x = percentage_heavy_drinkers, 
    y = life_expectancy_at_50,
    color=income_group
)) +
geom_point() +
facet_wrap(~ region, ncol=2)
```

:::

:::

:::::

> 
> So where does faceting fall short? For one, presenting each group on a separate chart makes it difficult to estimate the overlap between distributions. Its also harder to compare individual values across facets. 

Reviewer: ```suggestion
```

I moved this earlier
